Sources/MessagingPush/Util/DeepLinkUtil.swift

```diff
@@ -0,0 +1,21 @@
+import Common
+import Foundation
+
+protocol DeepLinkUtil: AutoMockable {
+    func isLinkValidNSUserActivityLink(_ url: URL) -> Bool
+}
+
+// sourcery: InjectRegister = "DeepLinkUtil"
+class DeepLinkUtilImpl: DeepLinkUtil {
+    // When using `NSUserActivity.webpageURL`, only certain URL schemes are allowed. An exception will be thrown otherwise which is why we have this function.
+    func isLinkValidNSUserActivityLink(_ url: URL) -> Bool {
```

Comment:
maybe I don't understand the problem statement correctly, here for which i apologize. 

But from what I understand, 

```
// to ger a URL from string, the method that gets used is 
URL(string: stringUrl)
```

and this already validates if it's a valid URL, that is it has the `allowedSchemes`

```
if let url = URL(string: stringUrl) {
    print(url)
} else {
    print("Invalid URL")
}
```

and the method, `isLinkValidNSUserActivityLink` is expecting an URL as well, so wouldn't this be already vetted? otherwise, the URL would always be `nil` ?

Reply:
I wonder if a class or function rename would help better understand the problem. 

This function's job is not to test if a URL is valid. The function's job is to verify that a URL is using only `http` or `https` for the scheme part of the URL string. 

This is needed because `MessagingPush+AppDelegate` file's code changes are using `NSUserActivity.webpageURL` and an exception is thrown if [a URL is using the wrong scheme](https://developer.apple.com/documentation/foundation/nsuseractivity/1418086-webpageurl). 

Do the tests for this class help explain the use of this function? 

Would a better comment help? Renaming of this function?

Reply:
Thanks, for the explanation, the [fixed test](https://github.com/customerio/customerio-ios/pull/265/commits/f68e813f7846d96b1b2e10a65192a1bc07f4b9a1) removed the confusion. 

A class renaming might help here because from my understanding there are two types of links, `deeplink` and `universal link`. And I was unsure where `NSUserActivity` lies in all of this. 

But i can't think of any other name either, so I am okay if you go with this or if you are able to think of another one.

Reply:
Deep links is an umbrella term. A deep link could be a universal link or an app scheme link. 

So, it's not deep link *or* universal link as universal links are deep links. I did add more details to deep links docs [in this pr](https://github.com/customerio/docs/pull/938) might be helpful for you to read those changes.